tools/setup.py

```diff
@@ -21,9 +22,16 @@ def main():
 
     write_lastchange()
 
-    mode = build_mode(default=None)
-    if mode is not None:
-        gn_gen(mode)
+    if "DENO_BUILD_PATH" in os.environ:
```

Comment:
After this patch, we no longer read `DENO_BUILD_MODE` env var.
So, I just add command line option here as well.
If we have DENO_BUILD_PATH and pass no option, runs gn_gen() with debug.
If we don't have DENO_BUILD_PATH and pass no option, runs gn_gen() with both of release and debug.
Is it acceptable? (I thought that this is similar to previous behavior)

Reply:
I think if someone specifies DENO_BUILD_PATH, that should be passed directly thru to gn_gen

Reply:
Sorry, I don't understand your comment exactly. (lack of my experience for deno)
As far as I remember, you wanted to generate GN files for both of debug and release.
The debug mode's default build path is `out/debug` and release mode's default build path is `out/release`.
But if someone specifies DENO_BUILD_PATH, then build path is the same in debug and release mode.
Then, which do you prefer to debug or release?

Reply:
I mean do something like this
```patch
From 36f21875187378df1fd04a553286cd5e817aed0e Mon Sep 17 00:00:00 2001
From: Ryan Dahl <ry@tinyclouds.org>
Date: Wed, 17 Oct 2018 09:30:59 -0400
Subject: [PATCH] x

---
 tools/setup.py | 16 ++++------------
 1 file changed, 4 insertions(+), 12 deletions(-)

diff --git a/tools/setup.py b/tools/setup.py
index 791c26e..fced7e8 100755
--- a/tools/setup.py
+++ b/tools/setup.py
@@ -23,18 +23,10 @@ def main():
     write_lastchange()
 
     if "DENO_BUILD_PATH" in os.environ:
-        parser = argparse.ArgumentParser()
-        setup_option_group = parser.add_mutually_exclusive_group()
-        setup_option_group.add_argument("--debug", action="store_true")
-        setup_option_group.add_argument("--release", action="store_true")
-        setup_option, _ = parser.parse_known_args()
-        if setup_option.release:
-            gn_gen("release")
-        else:
-            gn_gen("debug")
+        gn_gen()
     else:
-        gn_gen("release")
-        gn_gen("debug")
+        gn_gen(build_path("release"))
+        gn_gen(build_path("debug"))
 
 
 def write_if_not_exists(filename, contents):
@@ -143,7 +135,7 @@ def generate_gn_args(mode):
 
 
 # gn gen.
-def gn_gen(mode):
+def gn_gen(mode=None):
     # Rather than using gn gen --args we write directly to the args.gn file.
     # This is to avoid quoting/escaping complications when passing overrides as
     # command-line arguments.
-- 
2.15.0

```

Reply:
DENO_BUILD_PATH is in case you want a custom build... say out/my_custom_build

each build path can have different build parameters.

Reply:
Sorry, I might still be missing something..

Do you want to only allow handcraft if DENO_BUILD_PATH is set?
In case of auto-generation, the gn_gen() calls generate_gn_args() internally. It only allows release or debug.
```python
def generate_gn_args(mode):
    out = []
    if mode == "release":
        out += ["is_official_build=true"]
    elif mode == "debug":
        # Enable Jumbo build by default in debug mode for faster build.
        # https://chromium.googlesource.com/chromium/src/+/master/docs/jumbo.md
        out += ["use_jumbo_build=true"]
    else:
        print "Bad mode {}. Use 'release' or 'debug' (default)" % mode
        sys.exit(1)
```

Before this patch, we could set build mode. We can see this use-case in .travis.xml.
```
export DENO_BUILD_PATH="out/Default"
export DENO_BUILD_MODE="release"
```
So, the `out/hello/args.gn` could be generated as `release` mode.
If not set DENO_BUILD_MODE, it will be generated as `debug` mode.

If DENO_BUILD_PATH is set and we don't allow additional parameter, how can we generate `out/hello/args.gn` to `release` mode (in case of auto-generation instead of handcraft)?

Reply:
> Do you want to only allow handcraft if DENO_BUILD_PATH is set?

I thought about it more... so setup should always create out/release and out/debug... If DENO_BUILD_PATH that shouldn't change the behavior of setup.py. We will leave it to the user to have generated their DENO_BUILD_PATH...
In the case of build.py - if DENO_BUILD_PATH is set it should use that.